library/core/src/lib.rs

```diff
@@ -233,6 +235,44 @@ use prelude::v1::*;
 #[macro_use]
 mod macros;
 
+#[allow(missing_docs, missing_debug_implementations)]
+#[unstable(feature = "autoref", issue = "none")]
+pub mod autoref {
+    // See below how this is used as a const-generic parameter.
+    #[unstable(feature = "autoref", issue = "none")]
+    #[derive(Eq, PartialEq)]
+    pub struct UnstableMethodSeal;
+
+    #[unstable(feature = "autoref", issue = "none")]
+    pub trait AutoRef {
+        // Not the best name but trying to avoid conflicts as much as possible.
+        #[unstable(feature = "autoref", issue = "none")]
+        #[inline(always)]
+        fn __rustc_unstable_auto_ref_mut_helper<
+            // This is the clever bit: no stable method could have this parameter:
+            const _SEAL: UnstableMethodSeal,
+        >(
+            &mut self,
+        ) -> &mut Self {
+            self
+        }
+    }
+
+    #[unstable(feature = "autoref", issue = "none")]
+    impl<T: ?Sized> AutoRef for T {}
+
+    // Appropriate allow_internal_unstable attributes go here:
```

Comment:
Heh, all 4 of my comments are a bit off-topic, they were for the demonstration.
Also, I'm not sure I would put it in `core::autoref` - maybe deeper in `core::borrow` or `core::intrinsics` or something?